upb/decode_fast.c

```diff
@@ -0,0 +1,330 @@
+      uint64_t hasbits, uint64_t data
+
+#define UPB_PARSE_ARGS d, ptr, msg, table, hasbits, data
+
+typedef enum {
+  CARD_s = 0,
+  CARD_o = 1,
+  CARD_r = 2,
+  CARD_p = 3
+} upb_card;
+
+UPB_NOINLINE
+const char *fastdecode_dispatch(upb_decstate *d, const char *ptr, upb_msg *msg,
+                                const upb_msglayout *table, uint64_t hasbits) {
+  uint16_t tag;
+  uint64_t data = 0;;
+  size_t idx;
+  if (UPB_UNLIKELY(ptr >= d->fastlimit)) {
+    return fastdecode_generic(UPB_PARSE_ARGS);
+  }
+  memcpy(&tag, ptr, 2);
+  idx = (tag & 0xf8) >> 3;
+  data = table->field_data[idx] ^ tag;
+  return table->field_parser[idx](UPB_PARSE_ARGS);
+}
+
+UPB_FORCEINLINE
+static bool fastdecode_checktag(uint64_t data, int tagbytes) {
+  if (tagbytes == 1) {
+    return (data & 0xff) == 0;
+  } else {
+    return (data & 0xffff) == 0;
+  }
+}
+
+UPB_FORCEINLINE uint64_t fastdecode_munge(uint64_t val, int valbytes, bool zigzag) {
+  if (valbytes == 1) {
+    return val != 0;
+  } else if (zigzag) {
+    if (valbytes == 4) {
+      uint32_t n = val;
+      return (n >> 1) ^ -(int32_t)(n & 1);
+    } else if (valbytes == 8) {
+      return (val >> 1) ^ -(int64_t)(val & 1);
+    }
+    UPB_UNREACHABLE();
+  }
+  return val;
+}
+
+UPB_FORCEINLINE
+static uint16_t fastdecode_readtag(const char *ptr, int tagbytes) {
+  uint16_t ret = 0;
+  memcpy(&ret, ptr, tagbytes);
+  return ret;
+}
+
+UPB_FORCEINLINE
+static void *fastdecode_getfield_ofs(upb_decstate *d, const char *ptr,
+                                     upb_msg *msg, size_t ofs, uint64_t *data,
+                                     uint64_t *hasbits, upb_array **outarr,
+                                     int tagbytes, int valbytes,
+                                     upb_card card) {
+  void *field = (char *)msg + ofs;
+
+  switch (card) {
+    case CARD_s:
+      *hasbits |= *data;
+      return field;
+    case CARD_o: {
+      uint32_t *case_ptr = UPB_PTR_AT(msg, (*data >> 16) & 0xffff, uint32_t);
+      *case_ptr = (*data >> 32) & 0xffff;
+      return field;
+    }
+    case CARD_r: {
+      uint8_t elem_size_lg2 = __builtin_ctz(valbytes);
+      upb_array **arr_p = field;
+      upb_array *arr;
+      uint64_t elem_avail;
+      uint16_t expected_tag;
+      *hasbits >>= 16;
+      *(uint32_t*)msg |= *hasbits;
+      *hasbits = 0;
+      if (UPB_LIKELY(!*arr_p)) {
+        size_t need = (valbytes * 4) + sizeof(upb_array);
+        if (UPB_UNLIKELY((size_t)(d->arena_end - d->arena_ptr) < need)) {
+          *data = 0;
+          return NULL;
+        }
+        arr = (void*)d->arena_ptr;
+        field = arr + 1;
+        arr->data = _upb_array_tagptr(field, elem_size_lg2);
+        *arr_p = arr;
+        arr->size = 4;
+        arr->len = 0;
+        elem_avail = 4;
+        d->arena_ptr += need;
+      } else {
+        arr = *arr_p;
+        field = _upb_array_ptr(arr);
+        elem_avail = arr->size - arr->len;
```

Comment:
Is there are a reason why we can't resize if elem_avail is 0

Reply:
We speculate that elem_avail isn't 0. I now have inline resizing inside the loop when it is 0.